python-spec/src/somacore/scene.py

```diff
@@ -0,0 +1,270 @@
+"""A particular implementation of a point cloud."""
+
+_GeometryDataFrame = TypeVar("_GeometryDataFrame", bound=spatialdata.GeometryDataFrame)
+"""A particular implementation of a geometry dataframe."""
+
+_RootSO = TypeVar("_RootSO", bound=base.SOMAObject)
+"""The root SomaObject type of the implementation."""
+
+
+class Scene(
+    collection.BaseCollection[_RootSO],
+    Generic[_MultiscaleImage, _PointCloud, _GeometryDataFrame, _RootSO],
+):
+    """A collection subtype representing spatial assets that can all be stored
+    on a single coordinate space.
+
+    Lifecycle: experimental
+    """
+
+    # This class is implemented as a mixin to be used with SOMA classes.
+    # For example, a SOMA implementation would look like this:
+    #
+    #     # This type-ignore comment will always be needed due to limitations
+    #     # of type annotations; it is (currently) expected.
+    #     class Scene(  # type: ignore[type-var]
+    #         ImplBaseCollection[ImplSOMAObject],
+    #         somacore.Scene[
+    #             ImplMultiscaleImage,
+    #             ImplPointCloud,
+    #             ImplGeometryDataFrame,
+    #             ImplSOMAObject,
+    #         ],
+    #     ):
+    #         ...
+
+    __slots__ = ()
+    soma_type: Final = "SOMAScene"  # type: ignore[misc]
+
+    img = _mixin.item[collection.Collection[_MultiscaleImage]]()
+    """A collection of multiscale images backing the spatial data.
+
+    Lifecycle: experimental
+    """
+
+    obsl = _mixin.item[collection.Collection[Union[_PointCloud, _GeometryDataFrame]]]()
+    """A collection of observation location data.
+
+    This collection exists to store any spatial data in the scene that joins on the obs
+    ``soma_joinid``. Each dataframe in ``obsl`` can be either a PointCloud
+    or a GeometryDataFrame.
+    
+    Lifecycle: experimental
+    """
+
+    varl = _mixin.item[
+        collection.Collection[
+            collection.Collection[Union[_PointCloud, _GeometryDataFrame]]
+        ]
+    ]()
+    """A collection of collections of variable location data.
+
+    This collection exists to store any spatial data in the scene that joins on the
+    variable ``soma_joinid`` for the measurements in the SOMA experiment. The top-level
+    collection maps from measurement name to a collection of dataframes.
+
+    Each dataframe in a ``varl`` subcollection can be either a GeometryDataFrame or a
+    PointCloud.
+    
+    Lifecycle: experimental
+    """
+
+    @property
+    @abc.abstractmethod
+    def coordinate_space(self) -> Optional[coordinates.CoordinateSpace]:
+        """Coordinate system for this scene.
+
+        Lifecycle: experimental
+        """
+        raise NotImplementedError()
+
+    @coordinate_space.setter
+    @abc.abstractmethod
+    def coordinate_space(self, value: coordinates.CoordinateSpace) -> None:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def register_geometry_dataframe(
```

Comment:
Soma collection objects like `Experiment` and `Measurement` have methods like `add_new_dataframe`. Why does this class have `register_` methods instead?

Reply:
This has those methods as well (they are inherited from the `BaseCollection`). These methods record a transform from the `Scene` to the assets.

Reply:
Could we name these methods using the same scheme used for the getters: `<op>_transformation_<class>()`?
```suggestion
    def set_transformation_geometry_dataframe(
```